### PR TITLE
refactor: refactor tests to use HTPipeTest function(new)

### PR DESCRIPTION
**Pull request by VolodymyrWork23**

### CODE REVIEW

File: test/index.test.ts [hunk 1]
```diff
@@ -14,6 +36,81 @@ interface MockUser {
...
+async function HTPipeTest<
+  TPipe extends OptionallyHasInitPreContext<any, any> &
+    OptionallyHasSanitizeParams<any, any> &
+    OptionallyHasSanitizeBody<any, any> &
+    MightHavePreAuthorize<any, any> &
+    OptionallyHasAttachData<any, any> &
+    MightHaveFinalAuthorize<any, any> &
+    OptionallyHasDoWork<any, any> &
+    MightHaveRespond<any, any> &
+    MightHaveSanitizeResponse<any, any>,
```
Comment (captaincaius) [inline test/index.test.ts:55]: This isn't quite what we discussed in chat...
- [ ] why is TPipeIn extends TPipeInExpected?
- [ ] why is TLifecycleStage dependent on what TPipe HAPPENS to extend in some order?  Shouldn't it just be based on TPipe and TStage?

File: test/index.test.ts [hunk 2]
```diff
@@ -14,6 +36,81 @@ interface MockUser {
...
+      : never
+    : never,
+  TPipeInExpected = TPipeIn,
+  TPipeOutExpected = TPipeOut
+>(
+  pipe: TPipe,
+  lifecycleStage: TStage,
+  pipeIn: TPipeIn,
+  pipeOut: TPipeOut,
+  valid: TValid,
```
Comment (captaincaius) [inline test/index.test.ts:94]: I don't think we need pipeInExpected and pipeOutExpected FUNCTION parameters ... we just needed the optional TYPE parameters earlier.  99% of the time, it won't even be used, and it'll just be TPipeIn and TPipeOut respectively
- [ ] remove function params.

File: test/index.test.ts [hunk 3]
```diff
@@ -14,6 +36,81 @@ interface MockUser {
...
+  pipeInExpected: TPipeInExpected = pipeIn,
+  pipeOutExpected: TPipeOutExpected = pipeOut
+) {
+  const pipedLifecycleStage = pipe[lifecycleStage];
+
```
Comment (captaincaius) [inline test/index.test.ts:99]: Can you explain this if/else?  I don't quite get it - isn't it a tautology that the else branch will always fail?  Or is there some truthiness edge case I'm not thinking of with the cases where we want things to be empty?

Comment (captaincaius) [reply test/index.test.ts:99]: - [ ] explain

to be a little clearer, on the surface it looks like this code can/should just be:
```
expect(pipedLifecycleStage).to.not.be.empty;
if(pipedLifecycleStage) {
...
}
```

File: test/index.test.ts (hunk 2 above)
Comment (captaincaius) [reply test/index.test.ts:94]: also, on the javascript side, can't you make valid into `valid: TValid = true` so that you don't have to pass it in each time?
- [ ] valid default to true

File: test/index.test.ts [hunk 4]
```diff
@@ -647,73 +678,17 @@ describe('HipThrusTS', () => {
...
+          { c: cReturned },
+          true
         );
-
-        expect(pipedRightOnlyAttachDataReturnValues).to.has.property('c');
-        expect(pipedRightOnlyAttachDataReturnValues.c).to.be.equal(cReturned);
       });
-      it('no attaches data when left and right is empty objects', () => {
+      it('no attaches data when left and right is empty objects', async () => {
         const pipedWithEmptyObjectsOnly = HTPipe({}, {});
```
Comment (captaincaius) [inline test/index.test.ts:691]: aren't we losing type coverage in this corner case?
- [ ] revive manual type coverage here

File: test/index.test.ts [hunk 5]
```diff
@@ -293,7 +390,116 @@ describe('HipThrusTS', () => {
     // @todo: add test coverage for TRIPLE-OPERATOR pipes
     // @todo: add test coverage for QUADRUPLE-AND-MORE-OPERATOR pipes (use empty objects)
     // @todo: add test coverage for multi-stage operators
```
Comment (captaincaius) [inline test/index.test.ts:392]: These are supposed to be the unit tests I was talking about right?
If so...
- [ ] simplify.  you're testing HTPipeTest itself - we don't need/want any left/right or even HTPipe.  Just send it an already built reference object. 
- [ ] You're testing types, but you should also test function.  Give HTPipe data with the CORRECT types but INCORRECT data to make sure it FAILS in that case.

For the type tests, I can't follow the strategy.  The things to think about are:
1. pipeIn/pipeOut
2. test for extra key exists on pipeIn (or pipeOut) compared to pipe
3. test for key missing on pipeIn (or pipeOut) compared to pipe
4. test for keys the same but type mismatch
So all-in-all I expect 2x3=6 tests for the types
- [ ] all 6 type failure cases

File: test/index.test.ts (hunk 1 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:55]: TPipeIn extends TPipeInExpected because of pipeInExpected param
now since i removed that param we don't need that anymore
other one done

File: test/index.test.ts (hunk 2 above)
Comment (VolodymyrWork23) [reply test/index.test.ts]: done
valid can't be true by default i tried that
the problem is it gives error when i try to set valid = true
error is:
Type 'true' is not assignable to type 'TValid'.
  'true' is assignable to the constraint of type 'TValid', but 'TValid' could be instantiated with a different subtype of constraint 'true'.

File: test/index.test.ts (hunk 3 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:115]: first of all you right and that else block is no needed
but if conditions is needed because without that condition i can't use the lifecycle() because it can be undefined

File: test/index.test.ts (hunk 4 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:781]: revived

File: test/index.test.ts (hunk 5 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:407]: done

File: test/index.test.ts (hunk 1 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:55]: just in case
TPipe extends HasInitPreContext<any, any>
      ? TPipe[TStage]
      : never
it was in () but prettier remove them

File: test/index.test.ts [hunk 6]
```diff
@@ -293,7 +405,191 @@ describe('HipThrusTS', () => {
...
+    describe('HTPipeTest', () => {
```
Comment (captaincaius) [inline test/index.test.ts:408]: I'd make attachData's return value actually use its input so we're sure there's no disconnection... so return something like `{c: cReturned, b: bReturned, aout: context.a }` instead
- [ ] validate wiring

File: test/index.test.ts [hunk 7]
```diff
@@ -293,7 +405,191 @@ describe('HipThrusTS', () => {
...
+            },
+            'attachData',
+            { a: aPassedIn },
+            { b: bReturned, c: incorrectCReturned },
+            true
+          );
+        } catch (err) {
+          errorReturned = err;
+        }
+        // tslint:disable-next-line:no-unused-expression
```
Comment (captaincaius) [inline test/index.test.ts:589]: are you sure this try/catch construct does what you expect??  For example, if you comment out all the expect() calls in HTPipeTest(), will this test fail?  I have a hunch that it won't.
- [ ] confirm this does what you think.

File: test/index.test.ts (hunk 2 above)
Comment (captaincaius) [reply test/index.test.ts]: bastard typescript

File: test/index.test.ts (hunk 6 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:408]: done

File: test/index.test.ts (hunk 7 above)
Comment (VolodymyrWork23) [reply test/index.test.ts:590]: yeah it's do what i want but it give wrong error message
so i decide to change it a little
expect(errorReturned).to.exist; will give error if errorReturned is undefined or null
